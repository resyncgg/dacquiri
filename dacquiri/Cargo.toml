[package]
name = "dacquiri"
version = "0.4.4"
edition = "2021"
authors = ["d0nut <d0nut@resync.gg>"]
description = "An authorization framework with compile-time enforcement."
repository = "https://github.com/resyncgg/dacquiri"
license = "MIT"
readme = "README.md"
homepage = "https://resync.gg"
keywords = ["authorization", "abac", "access", "authz", "security"]
categories = ["authentication"]

[features]
default = ["derive"]
derive = ["dacquiri_derive"]

[dependencies]
thiserror = "1"
async-trait = "0"
fxhash = "0"
dacquiri_derive = { version = "0.4", optional = true, path = "../dacquiri_derive" }

[dev-dependencies]
trybuild = "1"
paste = "1"
parking_lot = "0"
async-trait = "0"
rand = "0"
dacquiri_derive = { path = "../dacquiri_derive" }

[[example]]
name = "simple"

[[example]]
name = "bank"
