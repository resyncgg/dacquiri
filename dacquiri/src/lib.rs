#![deny(warnings)]
#![allow(incomplete_features)]
#![feature(adt_const_params)]
#![feature(associated_type_defaults)]
#![feature(generic_arg_infer)]
#![feature(generic_associated_types)]
#![feature(trait_alias)]
#![feature(marker_trait_attr)]
#![feature(negative_impls)]
#![feature(auto_traits)]
#![feature(rustc_attrs)]

#![doc = include_str!("../README.md")]

pub extern crate async_trait;

mod error;
mod attribute;
mod chain;
mod acquire;
mod has;
pub mod prelude;
mod constraint;
mod store;
mod private;

#[doc(hidden)]
pub const DEFAULT_ELEMENT_TAG: &str = "___";


