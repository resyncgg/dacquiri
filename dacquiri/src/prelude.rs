pub use crate::error::ConstraintError;
pub use crate::attribute::{
    BaseAttribute,
    AsyncAttribute,
    SyncAttribute,
    AttributeResult
};
pub use crate::chain::{
    ConstraintChain,
    ConstraintEntity,
    ConstraintResult,
    ConstraintT,
    EntityTag
};
pub use crate::store::ConstraintStore;
pub use crate::has::{
    HasConstraint,
    HasEntity,
    HasEntityWithType,
    ShedNext
};
pub use crate::acquire::{
    acquire_default::AcquireAttribute,
    acquire_with_resource::AcquireAttributeWithResource,
    acquire_with_context::AcquireAttributeWithContext,
    acquire_with_resource_and_context::AcquireAttributeWithResourceAndContext
};
pub use crate::constraint::InitializeConstraint;
pub use crate::store::EntityProof;

#[doc(hidden)]
pub use crate::DEFAULT_ELEMENT_TAG;

#[cfg(feature = "derive")]
pub use dacquiri_derive::{
    policy,
    attribute
};