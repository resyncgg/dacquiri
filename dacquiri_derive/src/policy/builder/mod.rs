mod builder;
mod branch;
mod clause;

pub use builder::PolicyBuilder;