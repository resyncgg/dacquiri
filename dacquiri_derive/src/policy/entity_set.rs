use std::collections::HashSet;

/// Refers to the usage of an entity in a context, clause, or other policy construct
#[derive(Clone, Hash, Eq, PartialEq)]
pub(crate) struct EntityRef(String);

impl From<String> for EntityRef {
    fn from(value: String) -> Self {
        Self(value)
    }
}

impl ToString for EntityRef {
    fn to_string(&self) -> String {
        self.0.clone()
    }
}

pub(crate) trait EntitySet {
    fn common_entities(&self) -> HashSet<EntityRef>;
}