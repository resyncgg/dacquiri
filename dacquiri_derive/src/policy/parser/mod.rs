mod policy;
mod entities;
pub(crate) mod context;
mod clauses;

pub(crate) use policy::Policy;
pub(crate) use entities::EntityDeclaration;
use crate::utils::NonstandardKeyword;

pub(crate) const ENTITIES_KEYWORD: &str = "entities";
pub(crate) const CONTEXT_KEYWORD: &str = "context";
pub(crate) const IS_KEYWORD: &str = "is";

pub(crate) type IsKeyword = NonstandardKeyword<IS_KEYWORD>;