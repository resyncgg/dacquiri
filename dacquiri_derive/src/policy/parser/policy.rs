use proc_macro2::Span;
use quote::ToTokens;
use syn::{
    Ident,
    Token,
};
use syn::parse::{Parse, ParseStream};

use super::{
    ENTITIES_KEYWORD,
    GUARD_KEYWORD,
    entities::Entities,
    guard::Guard,
};


/**
Example

#[policy(
    entities = (
        user: User,
        team_one: Team,
        team_two: Team,
    ),
    guard = (
        ActiveTeamMember(user, team_one),
        ActiveTeamMember(user, team_two)
    ),
    guard = (
        user is PlatformAdmin
    )
)]
pub trait MultiTeamMember {
    fn transfer_data(&self) {
        let team_ones_files = self.as_policy::<ActiveTeamMember<user, team_one>>().get_files(); // team one

        self.as_policy::<ActiveTeamMember<user, team_two>>().add_files(team_ones_files); // team two
    }
}

#[policy(
    entities = (
        user: User,
        team: Team
    ),
    guard = (
        user is EnabledUser,
        team is EnabledTeam,
        user is TeamMember for team,
    )
)]
pub trait ActiveTeamMember {
    fn get_files(&self) -> Vec<File> { /* .. */ }
    fn add_files(&self, files: Vec<File>) { /* .. */ }
}
 */

enum PolicyKey {
    Entities,
    Guard
}

impl TryFrom<Ident> for PolicyKey {
    type Error = ();

    fn try_from(value: Ident) -> Result<Self, Self::Error> {
        match value.to_token_stream().to_string() {
            token if token == ENTITIES_KEYWORD => Ok(PolicyKey::Entities),
            token if token == GUARD_KEYWORD => Ok(PolicyKey::Guard),
            _ => Err(())
        }
    }
}

impl Parse for PolicyKey {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let value: Ident = input.parse()?;

        value
            .try_into()
            .map_err(|_| {
                syn::Error::new(Span::call_site(), "Keyword was invalid.")
            })
    }
}

pub(crate) struct Policy {
    pub(crate) entities: Entities,
    pub(crate) guards: Vec<Guard>
}

impl Parse for Policy {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let mut maybe_entities: Option<Entities> = None;
        let mut guards: Vec<Guard> = Vec::new();

        while let Ok(policy_key) = input.parse() {
            let _ = input.parse::<Token![=]>()?;

            match policy_key {
                PolicyKey::Entities => maybe_entities = Some(input.parse()?),
                PolicyKey::Guard => guards.push(input.parse()?),
            }

            if input.peek(Token![,]) {
                let _ = input.parse::<Token![,]>()?;
            }
        }

        let entities = maybe_entities
            .ok_or_else(|| syn::Error::new(
                Span::call_site(),
                "Entities are required to be defined for a policy."
            ))?;

        let policy = Policy {
            entities,
            guards
        };

        Ok(policy)
    }
}