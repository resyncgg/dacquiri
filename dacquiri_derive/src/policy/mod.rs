mod builder;
mod parser;
mod entity_set;

use syn::{ItemTrait, parse_macro_input};
use proc_macro::TokenStream;
use quote::ToTokens;
use crate::policy::builder::PolicyBuilder;
use crate::policy::parser::Policy;

pub(crate) fn handle_policy(args: TokenStream, input: TokenStream) -> TokenStream {
    let derive_input = parse_macro_input!(input as ItemTrait);
    let policy = parse_macro_input!(args as Policy);

    let policy_builder: PolicyBuilder = (policy, derive_input)
        .try_into()
        .expect("Invalid input.");

    policy_builder
        .to_token_stream()
        .into()
}

