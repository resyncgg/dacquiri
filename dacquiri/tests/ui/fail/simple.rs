#![deny(warnings)]
#![allow(incomplete_features)]
#![feature(generic_associated_types)]
#![feature(adt_const_params)]
#![feature(generic_arg_infer)]
#![feature(rustc_attrs)]
#![feature(marker_trait_attr)]

use dacquiri::prelude::*;
use policy::*;


fn main() -> AttributeResult<()> {
    let user = User::new(true, true);

    user.into_entity::<"user">()
        .prove::<Enabled<_, _>, "user">()?
        .print_message();

    Ok(())
}

mod policy {
    use dacquiri::prelude::*;

    pub struct User {
        enabled: bool,
        verified: bool
    }

    impl User {
        pub fn new(enabled: bool, verified: bool) -> Self {
            Self { enabled, verified }
        }
    }

    #[attribute(Enabled)]
    mod enabled {
        use super::User;

        #[attribute]
        pub fn check_enabled(user: &User) -> AttributeResult<()> {
            if user.enabled {
                Ok(())
            } else {
                Err(())
            }
        }
    }

    #[attribute(Verified)]
    mod verified {
        use super::User;

        #[attribute]
        pub fn check_verified(user: &User) -> AttributeResult<()> {
            if user.verified {
                Ok(())
            } else {
                Err(())
            }
        }
    }

    #[policy(
        entities = (
            user: User
        ),
        guard = (
            user is Enabled,
            user is Verified,
        )
    )]
    pub trait VerifiedUserPolicy {
        fn print_message(&self) {
            println!("Success!");
        }
    }
}
