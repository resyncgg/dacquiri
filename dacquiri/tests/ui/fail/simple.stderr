error[E0599]: the method `print_message` exists for struct `ConstraintChain<"user", "___", Enabled<policy::User, ()>, dacquiri::prelude::EntityProof<"user", policy::User, ConstraintStore>>`, but its trait bounds were not satisfied
  --> tests/ui/fail/simple.rs:18:10
   |
18 |         .print_message();
   |          ^^^^^^^^^^^^^ method cannot be called on `ConstraintChain<"user", "___", Enabled<policy::User, ()>, dacquiri::prelude::EntityProof<"user", policy::User, ConstraintStore>>` due to unsatisfied trait bounds
   |
  ::: src/chain.rs
   |
   | pub struct ConstraintChain<const STAG: EntityTag, const RTAG: EntityTag, Attr, Next> {
   | ------------------------------------------------------------------------------------
   | |
   | doesn't satisfy `_: VerifiedUserPolicyMarker<_>`
   | doesn't satisfy `_: policy::VerifiedUserPolicy<_>`
   |
note: trait bound `ConstraintChain<"user", "___", Enabled<policy::User, ()>, dacquiri::prelude::EntityProof<"user", policy::User, ConstraintStore>>: VerifiedUserPolicyMarker<_>` was not satisfied
  --> tests/ui/fail/simple.rs:65:5
   |
65 |       #[policy(
   |  _____^
   | |_____|
   | |
66 | |         entities = (
67 | |             user: User
68 | |         ),
...  |
72 | |         )
73 | |     )]
   | |______^
   | |______|
   |
   = note: this error originates in the attribute macro `policy` (in Nightly builds, run with -Z macro-backtrace for more info)
